File: ruby/beer-song/beer_song.rb
class BeerSong
  def self.recite(c, n_verses)
    verses = []

    penultimate_line = <<~EOM
    1 bottle of beer on the wall, 1 bottle of beer.
    Take it down and pass it around, no more bottles of beer on the wall.
    EOM

    last_line = <<~EOM
    No more bottles of beer on the wall, no more bottles of beer.
    Go to the store and buy some more, 99 bottles of beer on the wall.
    EOM

    if c == 0
      last_line
    else
      while c > 1 do
        verses << <<~EOM
        #{c} bottles of beer on the wall, #{c} bottles of beer.
        Take one down and pass it around, #{c-1} bottle#{c <= 2 ? nil : 's'} of beer on the wall.
        EOM
        c -= 1
      end
      verses.push(penultimate_line, last_line)
      verses.first(n_verses).join("\n")
    end

  end
end

File: javascript/beer-song/beer-song.js
function bottles(number) {
  switch (number) {
    case 0:
      return 'No more bottles';
    case 1:
      return '1 bottle';
    default:
      return `${number} bottles`;
  }
}

function action(verseN) {
  if (verseN === 0) {
    return 'Go to the store and buy some more, ';
  }

  return `Take ${verseN === 1 ? 'it' : 'one'} down and pass it around, `;
}

function nextBottle(verseN) {
  return `${bottles(verseN === 0 ? 99 : verseN - 1).toLowerCase()} of beer on the wall.\n`;
}

class BeerSong {
  static verse(number) {
    const line1 = `${bottles(number)} of beer on the wall, `;
    const line2 = `${bottles(number).toLowerCase()} of beer.\n`;
    const line3 = action(number);
    const line4 = nextBottle(number);

    return [line1, line2, line3, line4].join('');
  }

  static sing(first = 99, last = 0) {
    const verses = [];
    for (let i = first; i >= last; i -= 1) {
      verses.push(this.verse(i));
    }

    return verses.join('\n');
  }
}

export default BeerSong;
